homeassistant/components/tplink/light.py

```diff
@@ -31,17 +35,31 @@
                     'convert to use the tplink component.')
 
 
-async def async_setup_entry(hass, config_entry, async_add_entities):
+async def async_setup_entry(
+        hass: HomeAssistantType,
+        config_entry,
+        async_add_entities
+):
     """Set up discovered switches."""
-    devs = []
-    for dev in hass.data[TPLINK_DOMAIN][CONF_LIGHT]:
-        devs.append(TPLinkSmartBulb(dev))
-
-    async_add_entities(devs, True)
+    async_add_entities_retry(
+        hass,
+        async_add_entities,
+        hass.data[TPLINK_DOMAIN][CONF_LIGHT],
+        add_entity
+    )
 
     return True
 
 
+def add_entity(device: SmartBulb, async_add_entities):
+    """Check if device is online and add the entity."""
+    device.get_sysinfo()
```

Comment:
There is no need for this, as `get_features` is called during the update, which gets called before the entity is added due to `update_before_add`.

Reply:
The goal here is to only add entities to HA that we know will return sysinfo (online devices). By adding get_sysinfo before the add_entities call, we have the opportunity for async_add_entities_retry to catch the resulting error and attempt to add the entity later. This is necessary for devices that are not online during HA boot.

Without this, the code will behave as it did before where static devices crash the component during HA boot because it cannot pull values from sysinfo and at the very least not have a unique_id.

Reply:
Ok, maybe just add a simple commentary on why it's done here instead of deferred to its right place.